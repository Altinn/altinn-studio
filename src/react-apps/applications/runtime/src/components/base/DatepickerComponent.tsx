import * as React from 'react';
import '../../styles/InputComponent.css';

export interface IDatePickerProps {
  id: string;
  component: IFormComponent;
  formData: any;
  handleDataChange: (value: any) => void;
  isValid?: boolean;
}

export interface IDatePickerState {
  value: string;
}

export class DatepickerComponent
  extends React.Component<IDatePickerProps, IDatePickerState> {

  constructor(_props: IDatePickerProps, _state: IDatePickerState) {
    super(_props, _state);
    this.state = {
      value: _props.formData ? _props.formData : '',
    };
  }

  public onDataChanged = (e: any) => {
    this.setState({
      value: e.target.value,
    });
  }

  public onDataChangeSubmit = () => {
    this.props.handleDataChange(this.state.value);
  }

  public componentDidMount() {
    (window as any).initDatePicker();
  }

  public render() {
    return (
      <div className='form-group a-form-group a-form-group-datepicker'>
        <div className='input-group'>
          <input
            type='text'
            id={this.props.id}
            className={this.props.isValid ?
              'form-control a-hasButton date' :
              'form-control a-hasButton date validation-error'}
            onBlur={this.onDataChangeSubmit}
            onChange={this.onDataChanged}
            disabled={this.props.component.disabled}
            required={this.props.component.required}
            value={this.state.value}
          />
          <div className='input-group-prepend a-icon-right'>
            <i className='ai ai-date' />
          </div>
        </div>
      </div>
    );
  }
}
