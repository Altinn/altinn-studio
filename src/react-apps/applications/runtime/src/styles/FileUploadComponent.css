.cursor-pointer{
  cursor: pointer;
}

.file-upload-text-bold {
  font-weight: 400;
  font-size: 1.6rem;
  margin-bottom: 0.6rem;
}

.file-upload-text-bold-small {
  font-weight: 500;
  font-size: 1.2rem;
  margin-bottom: 0;
}

.file-upload-text {
  font-weight: 300;
  font-size: 1.6rem;
}

.file-upload-invalid:focus {
  outline-color: #E23B53;
}

.blue-underline {
  border-bottom: 2px solid #1EADF7;
}

.blue-underline-dotted {
  border-bottom: 1px dotted #1EADF7;
}

.file-upload-table > thead > tr > th {
  font-size: 1.4rem;
  font-weight: 500 !important;
}

.file-upload-table > tbody > tr > td {
  padding-top: 1.2rem;
  padding-bottom: 1.2rem;
}

.file-upload-table > tbody > tr > td:last-child > div {
  cursor: pointer;
  float: right;
  padding-right: 0;
}

.file-upload-table > tbody > tr > td:last-child {
  padding-right: 0;
  width: 15%;
}

.file-upload-table > tbody > tr:last-child{
  border: 0;
}

.file-upload-table {
  width: 100%;
  margin-top: 2.4rem;
  border-spacing: 0 1.2rem;
}

.file-upload-table > tbody {
  font-weight: 500 !important;
  font-size: 1.2rem;
}

.file-upload-button {
  cursor: pointer;
  background: 0;
  font-size: 1.6rem;
  border-left: 0;
  border-right: 0;
  border-top: 0;
  padding: 0;
  margin-top: 2.4rem;
}
