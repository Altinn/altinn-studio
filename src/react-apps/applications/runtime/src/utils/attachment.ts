import { IAttachmentListApiResponse, IAttachments } from '../shared/resources/attachments';

export function mapAttachmentListApiResponseToAttachments(response: IAttachmentListApiResponse[]): IAttachments {
  const attachments: IAttachments = {};
  if (!response) {
    return attachments;
  }
  response.forEach((attachmentsByType: IAttachmentListApiResponse) => {
    if (!attachmentsByType.type) {
      return;
    }
    attachments[attachmentsByType.type] = [];
    if (!attachmentsByType.attachments) {
      return;
    }
    attachmentsByType.attachments.forEach((attachment) => {
      attachments[attachmentsByType.type].push(
        { ...attachment, uploaded: true, deleting: false });
    });
  });
  return attachments;
}
