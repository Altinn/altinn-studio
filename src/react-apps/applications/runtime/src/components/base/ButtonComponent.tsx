import * as React from 'react';
import { connect } from 'react-redux';
import { makeGetValidationsSelector } from '../../selectors/getValidations';
import { IValidations } from '../../types/global';
import { canFormBeSaved, getErrorCount } from '../../utils/validation';
import FormDataActions from './../../features/form/data/actions/index';
import { IAltinnWindow, IRuntimeState } from './../../types';
export interface IButtonProvidedProps {
  id: string;
  text: string;
  disabled: boolean;
  handleDataChange: (value: any) => void;
  unsavedChanges: boolean;
  formDataCount: number;
}

export interface IButtonProps extends IButtonProvidedProps {
  validations: IValidations;
}

export interface IButtonState { }

export class ButtonComponentClass extends React.Component<IButtonProps, IButtonState> {
  public renderSubmitButton = () => {
    const disabled = (getErrorCount(this.props.validations) > 0) || this.props.unsavedChanges;
    return (
      <button
        type='submit'
        className={disabled ? 'a-btn a-btn-success disabled' : 'a-btn a-btn-success'}
        onClick={this.submitForm}
        disabled={disabled}
        id={this.props.id}
        style={{ marginBottom: '0' }}
      >
        {this.props.text}
      </button>
    );
  }

  // TODO: Remove saveButton and functions (and sagas) when we have implemented automatic save.
  public renderSaveButton = () => {
    const disabled = !this.props.unsavedChanges || !canFormBeSaved(this.props.validations);
    return (
      <button
        type='submit'
        className={disabled ?
          'a-btn a-btn-success disabled' : 'a-btn a-btn-success'}
        onClick={this.saveFormData}
        disabled={disabled}
        id='saveBtn'
        style={{ marginBottom: '0' }}
      >
        Lagre
      </button>
    );
  }

  public saveFormData = () => {
    const { org, app, instanceId } = window as Window as IAltinnWindow;
    FormDataActions.submitFormData(
      `${window.location.origin}/${org}/${app}/api/${instanceId}`,
    );
  }

  public submitForm = () => {
    const {org, app, instanceId } = window as Window as IAltinnWindow;
    FormDataActions.submitFormData(
      `${window.location.origin}/${org}/${app}/api/${instanceId}`,
      'Complete',
    );
  }

  public render() {
    return (
      <div className='row'>
        <div className='a-btn-group' style={{ marginTop: '3.6rem', marginBottom: '0' }}>
          {this.renderSaveButton()}
          {this.renderSubmitButton()}
        </div>
      </div>
    );
  }
}

const makeMapStateToProps = () => {
  const GetValidations = makeGetValidationsSelector();
  const mapStateToProps = (state: IRuntimeState, props: IButtonProvidedProps): IButtonProps => {
    return {
      validations: GetValidations(state),
      ...props,
    };
  };
  return mapStateToProps;
};

export const ButtonComponent = connect(makeMapStateToProps)(ButtonComponentClass);
