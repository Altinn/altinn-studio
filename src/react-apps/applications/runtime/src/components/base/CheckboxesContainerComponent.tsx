import { FormControlLabel, FormGroup } from '@material-ui/core';
import Checkbox, { CheckboxProps } from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import { makeStyles } from '@material-ui/core/styles';
import classNames = require('classnames');
import * as React from 'react';
import altinnTheme from '../../../../shared/src/theme/altinnAppTheme';
import { renderValidationMessagesForComponent } from '../../utils/render';

export interface ICheckboxContainerProps {
  id: string;
  formData: any;
  handleDataChange: (value: any) => void;
  handleFocusUpdate: (value: any) => void;
  isValid: boolean;
  validationMessages: any;
  options: any[];
  preselectedOptionIndex: number;
  readOnly: boolean;
  shouldFocus: boolean;
}

const useStyles = makeStyles({
  root: {
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
  icon: {
    'border': '2px solid #1EAEF7',
    'width': 24,
    'height': 24,
    'backgroundColor': '#ffffff',
    '$root.Mui-focusVisible &': {
      outline: '2px solid #ff0000',
      outlineOffset: 0,
      outlineColor: altinnTheme.altinnPalette.primary.blueDark,
    },
    'input:hover ~ &': {
      borderColor: altinnTheme.altinnPalette.primary.blueDark,
    },
    'input:disabled ~ &': {
      boxShadow: 'none',
      background: 'rgba(206,217,224,.5)',
    },
  },
  checkedIcon: {
    'backgroundColor': '#ffffff',
    '&:before': {
      display: 'block',
      width: 20,
      height: 20,
      // tslint:disable-next-line: max-line-length
      backgroundImage: "url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'%3E%3Cpath fill='%23000000' d='M6.564.75l-3.59 3.612-1.538-1.55L0 4.26 2.974 7.25 8 2.193z'/%3E%3C/svg%3E\")",
      content: '""',
    },
    'input:hover ~ &': {
      borderColor: altinnTheme.altinnPalette.primary.blueDark,
    },
  },
});

export const CheckboxContainerComponent = (props: ICheckboxContainerProps) => {
  const classes = useStyles(props);

  const [selected, setSelected] = React.useState([]);
  const checkBoxesIsRow: boolean = (props.options.length <= 2);

  React.useEffect(() => {
    returnState();
  }, [props]);

  const returnState = () => {
    if (
      !props.formData &&
      props.preselectedOptionIndex &&
      props.options &&
      props.preselectedOptionIndex < props.options.length
    ) {
      const preSelected: string[] = [];
      preSelected[props.preselectedOptionIndex] = props.options[props.preselectedOptionIndex].value;
      setSelected(preSelected);
    } else {
      setSelected(props.formData ? props.formData.split(',') : []);
    }
  };

  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSelected = selected;

    if (newSelected[event.target.value] === event.target.name) {
      newSelected[event.target.value] = '';
    } else {
      newSelected[event.target.value] = event.target.name;
    }

    setSelected(newSelected);
    props.handleDataChange(selectedHasValues(newSelected) ? newSelected.join() : '');
  };

  const selectedHasValues = (select: string[]): boolean => {
    return select.some((element) => element !== '');
  };

  const isOptionSelected = (option: string) => {
    return selected.indexOf(option) > -1;
  };

  const StyledCheckbox = (checkboxProps: CheckboxProps) => {
    return (
      <Checkbox
        className={classes.root}
        disableRipple={true}
        color='default'
        checkedIcon={<span className={classNames(classes.icon, classes.checkedIcon)} />}
        icon={<span className={classes.icon} />}
        inputProps={{ 'aria-label': 'decorative checkbox' }}
        {...checkboxProps}
      />
    );
  };

  return(
    <FormControl>
      <FormGroup row={checkBoxesIsRow}>
        {props.options.map((option, index) => (
          <React.Fragment key={index}>
            <FormControlLabel
              key={index}
              control={(
                <StyledCheckbox
                  checked={isOptionSelected(option.value)}
                  onChange={onDataChanged}
                  value={index}
                  name={option.value}
                />
              )}
              label={option.label}
            />
            { props.validationMessages &&
              this.isOptionSelected(option.value) &&
              renderValidationMessagesForComponent(props.validationMessages.simpleBinding, props.id) }
          </React.Fragment>
        ))}
      </FormGroup>
    </FormControl>
  );

};
