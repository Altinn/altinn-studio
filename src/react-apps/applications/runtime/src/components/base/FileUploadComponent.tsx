import * as React from 'react';
import DropZone from 'react-dropzone';
import { connect } from 'react-redux';
import uuid = require('uuid');
import altinnTheme from '../../../../shared/src/theme/altinnStudioTheme';
import { getLanguageFromKey } from '../../../../shared/src/utils/language';
import { IAttachment } from '../../sharedResources/attachments';
import AttachmentDispatcher from '../../sharedResources/attachments/attachmentActions';
import '../../styles/FileUploadComponent.css';
import { renderValidationMessagesForComponent } from '../../utils/render';

import { IRuntimeState } from '../../types';
import { IComponentValidations } from '../../types/global';

export interface IFileUploadProvidedProps {
  displayMode: string;
  hasCustomFileEndings?: boolean;
  id: string;
  isValid?: boolean;
  validationMessages?: IComponentValidations;
  language: any;
  maxFileSizeInMB: number;
  maxNumberOfAttachments: number;
  minNumberOfAttachments: number;
  readOnly: boolean;
  validFileEndings?: string;
}

export interface IFileUploadProps extends IFileUploadProvidedProps {
  attachments: IAttachment[];
}

export interface IFileUploadState {
  title: string;
  description: string;
  attachments: IAttachment[];
  showFileUpload: boolean;
  validations: string[];
  test: string;
}

const baseStyle = {
  width: 'auto',
  height: '15.6rem',
  borderWidth: '2px',
  borderColor: altinnTheme.altinnPalette.primary.blue,
  borderStyle: 'dashed',
  cursor: 'pointer',
};
const activeStyle = {
  borderStyle: 'solid',
};
const rejectStyle = {
  borderStyle: 'solid',
  borderColor: altinnTheme.altinnPalette.primary.red,
};
const validationErrorStyle = {
  borderStyle: 'dashed',
  borderColor: altinnTheme.altinnPalette.primary.red,
};

export const bytesInOneMB = 1048576;

export class FileUploadComponentClass
  extends React.Component<IFileUploadProps, IFileUploadState> {

  public static getDerivedStateFromProps(props: IFileUploadProps, state: IFileUploadState): IFileUploadState {
    return {
      ...state,
      attachments: props.attachments,
    };
  }

  constructor(props: IFileUploadProps, state: IFileUploadState) {
    super(props, state);
    this.state = {
      attachments: props.attachments || [],
      validations: [],
      ...state,
    };
  }

  public onDrop = (acceptedFiles: File[], rejectedFiles: File[]) => {
    const newFiles: IAttachment[] = [];
    const fileType = this.props.id; // component id used as filetype identifier for now, see issue #1364
    acceptedFiles.forEach((file: File) => {
      if ((this.state.attachments.length + newFiles.length) < this.props.maxNumberOfAttachments) {
        const tmpId: string = uuid();
        newFiles.push({ name: file.name, size: file.size, uploaded: false, id: tmpId, deleting: false });
        AttachmentDispatcher.uploadAttachment(file, fileType, tmpId, this.props.id);
      }
    });
    const validations: string[] = [];
    if (rejectedFiles.length > 0) {
      rejectedFiles.forEach((file) => {
        if (file.size > (this.props.maxFileSizeInMB * bytesInOneMB)) {
          validations.push(
            file.name + ' ' +
            getLanguageFromKey('form_filler.file_uploader_validation_error_file_size', this.props.language));
        } else {
          validations.push(
            getLanguageFromKey('form_filler.file_uploader_validation_error_general_1', this.props.language) + ' ' +
            file.name + ' ' +
            getLanguageFromKey('form_filler.file_uploader_validation_error_general_2', this.props.language));
        }
      });
    }
    const showFileUpload = (this.props.displayMode === 'simple') ? false :
      (this.props.attachments.length < this.props.maxFileSizeInMB);
    this.setState({
      attachments: this.state.attachments.concat(newFiles),
      // if simple mode, we should hide list on each drop
      showFileUpload,
      validations,
    });
  }

  public handleDeleteFile = (index: number) => {
    const attachmentToDelete = this.state.attachments[index];
    if (!attachmentToDelete.uploaded) {
      return;
    }
    const fileType = this.props.id; // component id used as filetype identifier for now, see issue #1364
    attachmentToDelete.deleting = true;
    const newList = this.state.attachments.slice();
    newList[index] = attachmentToDelete;
    this.setState({
      attachments: newList,
    });
    AttachmentDispatcher.deleteAttachment(attachmentToDelete, fileType, this.props.id);
  }

  public getComponentValidations = (): IComponentValidations => {
    const { validations } = this.state;
    let { validationMessages } = this.props;
    if (!validationMessages || !validationMessages.simpleBinding) {
      validationMessages = {
        ['simpleBinding']: {
          errors: [],
          warnings: [],
        },
      };
    }
    if (!validations || validations.length === 0) {
      return validationMessages;
    }
    validations.forEach((message) => {
      validationMessages.simpleBinding.errors.push(message);
    });
    return validationMessages;
  }

  public renderFileList = (): JSX.Element => {
    if (!this.state.attachments || this.state.attachments.length === 0) {
      return null;
    }
    return (
      <div id={'altinn-file-list-' + this.props.id}>
        <table className={'file-upload-table'}>
          <thead>
            <tr className={'blue-underline'} id={'altinn-file-list-row-header'}>
              <th>{getLanguageFromKey('form_filler.file_uploader_list_header_name', this.props.language)}</th>
              <th>{getLanguageFromKey('form_filler.file_uploader_list_header_file_size', this.props.language)}</th>
              <th>{getLanguageFromKey('form_filler.file_uploader_list_header_status', this.props.language)}</th>
              <th />
            </tr>
          </thead>
          <tbody>
            {this.state.attachments.map((attachment: IAttachment, index: number) => {
              return (
                <tr key={index} className={'blue-underline-dotted'} id={'altinn-file-list-row-' + attachment.id}>
                  <td>{attachment.name}</td>
                  <td>{attachment.size}</td>
                  <td>
                    {attachment.uploaded &&
                      <div>
                        {getLanguageFromKey('form_filler.file_uploader_list_status_done', this.props.language)}
                        < i className='ai ai-check-circle' />
                      </div>
                    }
                    {!attachment.uploaded &&
                      <div className='a-loader' id={'loader-upload'}>
                        <div
                          className='loader loader-ellipsis'
                          style={{ marginLeft: '1.3rem', marginBottom: '1.6rem' }}
                        />
                      </div>
                    }
                  </td>
                  <td>
                    <div onClick={this.handleDeleteFile.bind(this, index)} id={'attachment-delete-' + index}>
                      {!attachment.deleting &&
                        <>
                          {getLanguageFromKey('form_filler.file_uploader_list_delete', this.props.language)}
                          <i className='ai ai-trash' />
                        </>
                      }
                      {attachment.deleting &&
                        <div className='a-loader' id={'loader-delete'}>
                          <div
                            className='loader loader-ellipsis'
                            style={{ marginBottom: '1.6rem', marginRight: '1.0rem' }}
                          />
                        </div>
                      }
                    </div>
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
      </div>
    );
  }

  public renderFileUploadContent = (): JSX.Element => {
    return (
      <div className={'container'}>
        <div className='col text-center icon' style={{ marginTop: '3.5rem' }} >
          <i className='ai ai-upload' />
        </div>
        <div className='col text-center'>
          <p className={'file-upload-text-bold'}>
            {getLanguageFromKey('form_filler.file_uploader_drag', this.props.language)}
            <span className={'file-upload-text-bold blue-underline'}>
              {' ' + getLanguageFromKey('form_filler.file_uploader_find', this.props.language)}
            </span>
          </p>
        </div>
        <div className='col text-center'>
          <p className={'file-upload-text'}>
            {getLanguageFromKey('form_filler.file_uploader_valid_file_format', this.props.language)}
            {this.props.hasCustomFileEndings ? (' ' + this.props.validFileEndings) :
              (' ' + getLanguageFromKey('form_filler.file_upload_valid_file_format_all', this.props.language))}
          </p>
        </div>
      </div>
    );
  }

  public renderAddMoreAttachmentsButton = (): JSX.Element => {
    if (this.props.displayMode === 'simple' && !this.state.showFileUpload &&
      (this.state.attachments.length < this.props.maxNumberOfAttachments) && this.state.attachments.length > 0) {
      return (
        <button className={'file-upload-button blue-underline'} onClick={this.showFileUpload}>
          {getLanguageFromKey('form_filler.file_uploader_add_attachment', this.props.language)}
        </button>
      );
    } else {
      return null;
    }
  }

  public showFileUpload = () => {
    this.setState({
      showFileUpload: true,
    });
  }

  public shouldShowFileUpload = (): boolean => {
    const { attachments, showFileUpload } = this.state;
    if (attachments.length >= this.props.maxNumberOfAttachments) {
      return false;
    } else {
      return (this.props.displayMode !== 'simple') || (attachments.length === 0) ||
        (showFileUpload === true);
    }
  }

  public render() {
    const validationMessages = this.getComponentValidations();
    const showFileUpload: boolean = this.shouldShowFileUpload();
    const hasValidationMessages: boolean = validationMessages.simpleBinding.errors.length > 0;
    return (
      <div className={'container'} id={'altinn-fileuploader-' + this.props.id}>
        {showFileUpload &&
          <div>
            <p className={'file-upload-text-bold-small'}>
              {
                getLanguageFromKey('form_filler.file_uploader_max_size', this.props.language)
                + ' ' + this.props.maxFileSizeInMB + ' ' +
                getLanguageFromKey('form_filler.file_uploader_mb', this.props.language)
              }
            </p>
            <DropZone
              onDrop={this.onDrop}
              maxSize={this.props.maxFileSizeInMB * bytesInOneMB} // mb to bytes
              disabled={this.props.readOnly}
              accept={(this.props.hasCustomFileEndings) ? this.props.validFileEndings : null}
            >
              {({ getRootProps, getInputProps, isDragActive, isDragAccept, isDragReject }) => {
                let styles = { ...baseStyle };
                styles = isDragActive ? { ...styles, ...activeStyle } : styles;
                styles = isDragReject ? { ...styles, ...rejectStyle } : styles;
                styles = (hasValidationMessages) ? { ...styles, ...validationErrorStyle } : styles;

                return (
                  <div
                    {...getRootProps()}
                    style={styles}
                    id={'altinn-drop-zone-' + this.props.id}
                    className={'file-upload' + (hasValidationMessages ? ' file-upload-invalid' : '')}
                  >
                    <input {...getInputProps()} />
                    {this.renderFileUploadContent()}
                  </div>
                );
              }}
            </DropZone>
          </div>
        }
        {this.renderFileList()}
        {this.renderAddMoreAttachmentsButton()}
        <p className={'file-upload-text-bold-small'}>
          {
            getLanguageFromKey('form_filler.file_uploader_number_of_files', this.props.language) + ' ' +
            (this.props.minNumberOfAttachments ? this.state.attachments.length + '/' + this.props.minNumberOfAttachments
              : this.state.attachments.length)
          }
        </p>
        {(validationMessages.simpleBinding.errors.length > 0) &&
          renderValidationMessagesForComponent(validationMessages.simpleBinding,
            this.props.id)
        }
      </div>
    );
  }
}

const mapStateToProps = (state: IRuntimeState, props: IFileUploadProvidedProps): IFileUploadProps => {
  return {
    ...props,
    attachments: state.attachments.attachments[props.id] || [],
  };
};

export function getFileUploadComponentValidations(validationError: string, language: any): IComponentValidations {
  const componentValidations: any = {
    ['simpleBinding']: {
      errors: [],
      warnings: [],
    },
  };
  if (validationError === 'upload') {
    componentValidations.simpleBinding.errors.push(
      getLanguageFromKey('form_filler.file_uploader_validation_error_upload', language),
    );
  } else if (validationError === 'delete') {
    componentValidations.simpleBinding.errors.push(
      getLanguageFromKey('form_filler.file_uploader_validation_error_delete', language),
    );
  }
  return componentValidations;
}

export const FileUploadComponent = connect(mapStateToProps)(FileUploadComponentClass);
