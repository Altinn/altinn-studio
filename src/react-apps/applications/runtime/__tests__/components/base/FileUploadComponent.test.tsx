import { mount } from 'enzyme';
import 'jest';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { FileUploadComponent, FileUploadComponentClass, getFileUploadComponentValidations } from '../../../src/components/base/FileUploadComponent';
import { mapAttachmentListApiResponseToAttachments } from '../../../src/utils/attachment';

describe('>>> components/base/FileUploadComponent.tsx', () => {
  let mockDisplayMode: string;
  let mockHasCustomFileEndings: boolean;
  let mockId: string;
  let mockIsValid: boolean;
  let mockMaxFileSizeInMB: number;
  let mockMaxNumberOfAttachments: number;
  let mockMinNumberOfAttachments: number;
  let mockReadOnly: boolean;
  let mockAttachments: any[];
  let mockFileList: File[];
  let mockInitialState: any;
  let mockStore: any;

  beforeEach(() => {
    const createStore = configureStore();
    mockId = 'mockId';
    mockAttachments = [
      { name: 'attachment-name-1', id: 'attachment-id-1', size: '1200', uploaded: true, deleting: false },
      { name: 'attachment-name-2', id: 'attachment-id-2', size: '800', uploaded: false, deleting: false },
      { name: 'attachment-name-3', id: 'attachment-id-3', size: '400', uploaded: true, deleting: true },
    ];
    mockInitialState = {
      formAttachments: {
        attachments: {
          mockId: mockAttachments,
        },
        validationResults: {
          mockId: {
            simpleBinding: {
              errors: ['mock error message'],
            },
          },
        },
      },
    };
    mockMaxNumberOfAttachments = 4;
    mockMinNumberOfAttachments = 1;
    mockMaxFileSizeInMB = 2;
    mockHasCustomFileEndings = false;
    mockDisplayMode = 'simple';
    mockIsValid = true;
    mockReadOnly = false;
    mockFileList = [{ name: 'mock-name.txt', lastModified: null, size: 100, slice: null, type: null }];
    mockStore = createStore(mockInitialState);
  });

  it('+++ should match snapshot', () => {
    const rendered = renderer.create(
      <Provider store={mockStore}>
        <FileUploadComponent
          displayMode={mockDisplayMode}
          id={mockId}
          isValid={mockIsValid}
          language={{}}
          maxFileSizeInMB={mockMaxFileSizeInMB}
          maxNumberOfAttachments={mockMaxNumberOfAttachments}
          minNumberOfAttachments={mockMinNumberOfAttachments}
          readOnly={mockReadOnly}
        />
      </Provider>,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should show spinner when file is uploading or deleting', () => {
    const wrapper = mount(
      <FileUploadComponentClass
        displayMode={mockDisplayMode}
        id={mockId}
        isValid={mockIsValid}
        language={{}}
        maxFileSizeInMB={mockMaxFileSizeInMB}
        maxNumberOfAttachments={mockMaxNumberOfAttachments}
        minNumberOfAttachments={mockMinNumberOfAttachments}
        readOnly={mockReadOnly}
        attachments={mockAttachments}
      />,
    );
    expect(wrapper.find('#loader-upload')).toHaveLength(1);
    expect(wrapper.find('#loader-delete')).toHaveLength(1);
  });

  it('+++ should add validation error on onDrop rejection', () => {
    const wrapper = mount(
      <FileUploadComponentClass
        displayMode={mockDisplayMode}
        id={mockId}
        isValid={mockIsValid}
        language={{}}
        maxFileSizeInMB={mockMaxFileSizeInMB}
        maxNumberOfAttachments={mockMaxNumberOfAttachments}
        minNumberOfAttachments={mockMinNumberOfAttachments}
        readOnly={mockReadOnly}
        attachments={mockAttachments}
      />,
    );
    const instance = wrapper.instance() as FileUploadComponentClass;
    expect(instance.state.validations.length).toBe(0);
    instance.onDrop([], mockFileList);
    expect(instance.state.validations.length).toBe(1);
  });

  it('+++ should trigger onDelete on when delete is clicked and update state to deleting for that attachment', () => {
    const wrapper = mount(
      <FileUploadComponentClass
        displayMode={mockDisplayMode}
        id={mockId}
        isValid={mockIsValid}
        language={{}}
        maxFileSizeInMB={mockMaxFileSizeInMB}
        maxNumberOfAttachments={mockMaxNumberOfAttachments}
        minNumberOfAttachments={mockMinNumberOfAttachments}
        readOnly={mockReadOnly}
        attachments={mockAttachments}
      />,
    );
    const instance = wrapper.instance() as FileUploadComponentClass;
    const spy = jest.spyOn(instance, 'handleDeleteFile');
    wrapper.find('#attachment-delete-0').simulate('click');
    // workaround - have to click twice the first time
    wrapper.find('#attachment-delete-0').simulate('click');
    expect(instance.state.attachments[0].deleting).toBe(true);
    expect(spy).toHaveBeenCalled();
  });

  it('+++ should not display drop area when in simple mode and attachments exists', () => {
    const wrapper = mount(
      <FileUploadComponentClass
        displayMode={mockDisplayMode}
        id={mockId}
        isValid={mockIsValid}
        language={{}}
        maxFileSizeInMB={mockMaxFileSizeInMB}
        maxNumberOfAttachments={mockMaxNumberOfAttachments}
        minNumberOfAttachments={mockMinNumberOfAttachments}
        readOnly={mockReadOnly}
        attachments={mockAttachments}
      />,
    );
    expect(wrapper.find('#altinn-drop-zone-' + mockId)).toHaveLength(0);
  });

  it('+++ should map api-response to correct redux structure', () => {
    const mockApiResponse: any = [
      {
        type: 'mockType', attachments: [
          { name: 'name1', size: 400, id: 'id1' },
        ],
      },
      {
        type: 'mockType2', attachments: [
          { name: '2-name1', size: 400, id: '2-id1' },
          { name: '2-name2', size: 800, id: '2-id2' },
        ],
      },
      {
        type: 'mockType3', attachments: [],
      },
      {
        type: 'mockType4', // response with missing attachments
      },
      {
        attachments: [], // response with missing type
      },
      {
        // mapper should handle empty response
      },
    ];
    const mappedResponse = mapAttachmentListApiResponseToAttachments(mockApiResponse);

    expect(mappedResponse.mockType.length).toBe(1);
    expect(mappedResponse.mockType[0].name).toBe('name1');
    expect(mappedResponse.mockType[0].size).toBe(400);
    expect(mappedResponse.mockType[0].id).toBe('id1');

    expect(mappedResponse.mockType2.length).toBe(2);
    expect(mappedResponse.mockType2[0].name).toBe('2-name1');
    expect(mappedResponse.mockType2[0].size).toBe(400);
    expect(mappedResponse.mockType2[0].id).toBe('2-id1');
    expect(mappedResponse.mockType2[1].name).toBe('2-name2');
    expect(mappedResponse.mockType2[1].size).toBe(800);
    expect(mappedResponse.mockType2[1].id).toBe('2-id2');

    expect(mappedResponse.mockType3.length).toBe(0);

    expect(mappedResponse.mockType4.length).toBe(0);

  });
  it('+++ getFileUploadComponentValidations should return correct validation', () => {
    const mockLanguage = {
      language: {
        form_filler: {
          file_uploader_validation_error_delete: 'Noe gikk galt under slettingen av filen, prøv igjen senere.',
          file_uploader_validation_error_upload: 'Noe gikk galt under opplastingen av filen, prøv igjen senere',

        },
      },
    };
    let validation = getFileUploadComponentValidations('upload', mockLanguage.language);
    expect(validation).toEqual
      ({ simpleBinding: { errors: ['Noe gikk galt under opplastingen av filen, prøv igjen senere'], warnings: [] } });

    validation = getFileUploadComponentValidations('delete', mockLanguage.language);
    expect(validation).toEqual
      ({ simpleBinding: { errors: ['Noe gikk galt under slettingen av filen, prøv igjen senere.'], warnings: [] } });
  });

});
