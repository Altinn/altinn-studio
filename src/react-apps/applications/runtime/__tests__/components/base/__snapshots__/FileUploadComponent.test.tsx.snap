// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> components/base/FileUploadComponent.tsx +++ should match snapshot 1`] = `
<div
  className="container"
  id="altinn-fileuploader-mockId"
>
  <div
    id="altinn-file-list-mockId"
  >
    <table
      className="file-upload-table"
    >
      <thead>
        <tr
          className="blue-underline"
          id="altinn-file-list-row-header"
        >
          <th>
            form_filler.file_uploader_list_header_name
          </th>
          <th>
            form_filler.file_uploader_list_header_file_size
          </th>
          <th>
            form_filler.file_uploader_list_header_status
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr
          className="blue-underline-dotted"
          id="altinn-file-list-row-attachment-id-1"
        >
          <td>
            attachment-name-1
          </td>
          <td>
            1200
          </td>
          <td>
            <div>
              form_filler.file_uploader_list_status_done
              <i
                className="ai ai-check-circle"
              />
            </div>
          </td>
          <td>
            <div
              id="attachment-delete-0"
              onClick={[Function]}
            >
              form_filler.file_uploader_list_delete
              <i
                className="ai ai-trash"
              />
            </div>
          </td>
        </tr>
        <tr
          className="blue-underline-dotted"
          id="altinn-file-list-row-attachment-id-2"
        >
          <td>
            attachment-name-2
          </td>
          <td>
            800
          </td>
          <td>
            <div
              className="a-loader"
              id="loader-upload"
            >
              <div
                className="loader loader-ellipsis"
                style={
                  Object {
                    "marginBottom": "1.6rem",
                    "marginLeft": "1.3rem",
                  }
                }
              />
            </div>
          </td>
          <td>
            <div
              id="attachment-delete-1"
              onClick={[Function]}
            >
              form_filler.file_uploader_list_delete
              <i
                className="ai ai-trash"
              />
            </div>
          </td>
        </tr>
        <tr
          className="blue-underline-dotted"
          id="altinn-file-list-row-attachment-id-3"
        >
          <td>
            attachment-name-3
          </td>
          <td>
            400
          </td>
          <td>
            <div>
              form_filler.file_uploader_list_status_done
              <i
                className="ai ai-check-circle"
              />
            </div>
          </td>
          <td>
            <div
              id="attachment-delete-2"
              onClick={[Function]}
            >
              <div
                className="a-loader"
                id="loader-delete"
              >
                <div
                  className="loader loader-ellipsis"
                  style={
                    Object {
                      "marginBottom": "1.6rem",
                      "marginRight": "1.0rem",
                    }
                  }
                />
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    className="file-upload-button blue-underline"
    onClick={[Function]}
  >
    form_filler.file_uploader_add_attachment
  </button>
  <p
    className="file-upload-text-bold-small"
  >
    form_filler.file_uploader_number_of_files 3/1
  </p>
</div>
`;
